facilitator/src/aws_credentials.rs

```diff
@@ -1,160 +1,261 @@
+    /// Rusoto's mock credentials provider, wrapped in an Arc to provide
+    /// Send + Sync. Should only be used in tests.
+    Mock(Arc<MockCredentialsProvider>),
 }
 
-#[async_trait]
-impl ProvideAwsCredentials for DefaultCredentialsProvider {
-    async fn credentials(&self) -> Result<AwsCredentials, CredentialsError> {
-        self.0.credentials().await
+impl Provider {
+    /// Instantiates an appropriate Provider based on the provided configuration
+    /// values
+    pub fn new(identity: Identity, use_default_provider: bool) -> Result<Self> {
+        match (use_default_provider, identity) {
+            (true, _) => Self::new_default(),
+            (_, Some(identity)) => Self::new_web_identity_with_oidc(identity),
+            (_, None) => Self::new_web_identity_from_kubernetes_environment(),
+        }
     }
-}
-
-/// Provides AWS credentials from multiple possible sources using a priority order.
-///
-/// The following sources are checked in order for credentials when calling `credentials`:
-///
-/// 1. Environment variables: `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`
-/// 2. `credential_process` command in the AWS config file, usually located at `~/.aws/config`.
-/// 3. AWS credentials file. Usually located at `~/.aws/credentials`.
-/// 4. IAM instance profile. Will only work if running on an EC2 instance with an instance profile/role.
-///
-/// If the sources are exhausted without finding credentials, an error is returned.
-///
-/// The provider has a default timeout of 30 seconds. While it should work well for most setups,
-/// you can change the timeout using the `set_timeout` method.
-///
-/// # Example
-///
-///
-/// # Warning
-///
-/// This provider allows the [`credential_process`][credential_process] option in the AWS config
-/// file (`~/.aws/config`), a method of sourcing credentials from an external process. This can
-/// potentially be dangerous, so proceed with caution. Other credential providers should be
-/// preferred if at all possible. If using this option, you should make sure that the config file
-/// is as locked down as possible using security best practices for your operating system.
-///
-/// [credential_process]: https://docs.aws.amazon.com/cli/latest/topic/config-vars.html#sourcing-credentials-from-external-processes
-#[derive(Debug, Clone)]
-pub struct ChainProvider {
-    environment_provider: EnvironmentProvider,
-    instance_metadata_provider: InstanceMetadataProvider,
-    container_provider: ContainerProvider,
-    profile_provider: Option<ProfileProvider>,
-    webidp_provider: WebIdentityProvider,
-}
 
-impl ChainProvider {
-    /// Set the timeout on the provider to the specified duration.
-    #[allow(dead_code)]
-    pub fn set_timeout(&mut self, duration: Duration) {
-        self.instance_metadata_provider.set_timeout(duration);
-        self.container_provider.set_timeout(duration);
+    fn new_default() -> Result<Self> {
```

Comment:
Might be more rusty to use: https://doc.rust-lang.org/std/default/trait.Default.html

Reply:
The semantics are messy here: the default for this enum should probably be `new_web_identity_from_kubernetes_environment` to preserve backward compatibility, but `default` here refers to Rusoto's `DefaultCredentialsProvider`. This is one of those problems that I would classify as inherently stupid (we are forced by backward compat into making an API design that is bad), and thus any solution to it will reflect that stupidity.

Reply:
Love it! Sounds good to me. :shipit:

Reply:
The other thing I would note is that implementing `Default` is meant to be a convenience for the user of this enum, to spare them the hassle of filling in fields they don't care about or selecting an enum variant when they have no way of knowing which one they want. In this case, there is no such caller: we provide the `Provider::new()` associated function which does the right thing based on its inputs, so implementing `Default` wouldn't save anyone any hassle.